### PR TITLE
Extend ExecutionContext with states

### PR DESCRIPTION
**Pull request by erikzhang**
Closes #191

### CODE REVIEW

File: src/neo-vm/ExecutionContext.cs [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
 using System.Runtime.CompilerServices;
 
 namespace Neo.VM
 {
     [DebuggerDisplay("RVCount={RVCount}, InstructionPointer={InstructionPointer}")]
     public class ExecutionContext
     {
+        private readonly Dictionary<Type, object> states = new Dictionary<Type, object>();
```
Comment (shargon) [inline src/neo-vm/ExecutionContext.cs]: Maybe is better to use this https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject?view=netframework-4.8

Comment (erikzhang) [reply src/neo-vm/ExecutionContext.cs]: The performance of `ExpandoObject` is very poor.

**Comment by igormcoelho** [comment]
@erikzhang If I'm following your reasoning here, perhaps we should declare Context as sealed, and end future inheritance tries. This expresses more clearly the intend of not extending this.

File: src/neo-vm/ExecutionContext.cs (hunk 1 above)
Comment (igormcoelho) [reply src/neo-vm/ExecutionContext.cs]: I think that Map should resolve this for us Shargon... if performance is not the best, in time we improve it. Best thing now is to keep things moving.

File: src/neo-vm/ExecutionContext.cs [hunk 2]
```diff
@@ -1,11 +1,15 @@
-﻿using System.Diagnostics;
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
 using System.Runtime.CompilerServices;
 
 namespace Neo.VM
 {
     [DebuggerDisplay("RVCount={RVCount}, InstructionPointer={InstructionPointer}")]
     public class ExecutionContext
```
Comment (igormcoelho) [inline src/neo-vm/ExecutionContext.cs]:
```suggestion
    public sealed class ExecutionContext
```

Comment (igormcoelho) [reply src/neo-vm/ExecutionContext.cs]: This ends all future tries for extending this. Very clear intent.

File: src/neo-vm/ExecutionEngine.cs [hunk 3]
```diff
@@ -297,11 +294,11 @@ private bool ExecuteInstruction()
                                 context_pop.AltStack.CopyTo(CurrentContext.AltStack);
                             }
                             CheckStackSize(false, 0);
-                            ContextUnloaded?.Invoke(this, context_pop);
                             if (InvocationStack.Count == 0)
                             {
                                 State = VMState.HALT;
                             }
+                            ContextUnloaded(context_pop);
```
Comment (igormcoelho) [inline src/neo-vm/ExecutionEngine.cs]: So much better!

File: src/neo-vm/ExecutionEngine.cs [hunk 4]
```diff
@@ -1074,12 +1071,11 @@ private bool ExecuteInstruction()
             return true;
         }
 
-        private void LoadContext(ExecutionContext context)
+        protected virtual void LoadContext(ExecutionContext context)
```
Comment (igormcoelho) [inline src/neo-vm/ExecutionEngine.cs]: Will we still inherit from this?

Comment (shargon) [reply src/neo-vm/ExecutionEngine.cs]: Is the unique way to set the states

Comment (igormcoelho) [reply src/neo-vm/ExecutionEngine.cs]: So event based may be better than this, if we can seal the class and use event to manage internals...

**Comment by shargon** [comment]
How can I share the script Hash on ContextLoader? Also I want to cache this in this states, but i only can check if it is the same as the calling Context

**Comment by erikzhang** [comment]
What is ContextLoader?

**Comment by shargon** [comment]
sorry, `LoadContext` xD

**Comment by erikzhang** [comment]
You can override `LoadContext()`.

**Comment by shargon** [comment]
We should decide if is better with events or with `override`.
I think that events should have more presence in neo. But we can have both, are compatible

**Comment by erikzhang** [comment]
There are currently no other places where these events are used. And the event can be listened to by multiple listeners, and you can't control the order of execution, which is very unfavorable for the method that can modify the parameters.

**Comment by shargon** [comment]
Is ready to me, i am not sure if this could help with the script hash cache, because is related to the script hash, but this is not the priority now, just focus on https://github.com/neo-project/neo/issues/927

**Comment by igormcoelho** [comment]
@erikzhang we can still overide a method on sealed class? learning things every day...
I agree with you that execution order is fundamental, without that guarantee, its hard to use events :+1:

**Pull request merged**